testdata/p4_16_samples_outputs/gauntlet_hdr_assign_2-bmv2-midend.p4

```diff
@@ -36,17 +36,18 @@ control ingress(inout Headers h, inout Meta m, inout standard_metadata_t sm) {
     @hidden action gauntlet_hdr_assign_2bmv2l43() {
         h.h.setValid();
     }
-    @hidden action gauntlet_hdr_assign_2bmv2l38() {
+    @hidden action gauntlet_hdr_assign_2bmv2l37() {
+        h.h.a = 8w1;
```

Comment:
So the pass readds assignments that have no effect?

Reply:
The spec does not say that this assignment has no effect. Assigning to invalid header fields is really undefined, so having no effect is only one possible semantics, but this is a legal interpretation too. I have tweaked the meaning of setInvalid not to overwrite all fields of the header in the def-use analysis, that's why this assignment is no longer deleted.

Reply:
https://p4lang.github.io/p4-spec/docs/P4-16-v1.2.1.html#sec-uninitialized-values-and-writing-invalid-headers

Reply:
I am confused.  What sentence of the spec do you believe says that assigning to a field in a currently invalid header is "really undefined"?  I thought it was defined as having no effect upon any defined fields anywhere.

Reply:
From the section of the spec that you link above, in particular: "If any of these kinds of writes are performed:

+ a write to a field in a currently invalid header, either a regular header or an element of a header stack with an index that is in range, and that header is not part of a header_union
[ two bullet points irrelevant to my point omitted ]

then that write must not change any state that is currently defined in the system, neither in header fields nor anywhere else. In particular, if an invalid header is involved in the write, it must remain invalid."

Reply:
I think that the interpretation is that if you invalidate a header, write a field, and the make it valid again you *may* read back the value you wrote (or you may not). This is the semantics currently assumed by this analysis pass - it does not treat writes to invalid headers as "dead".

Reply:
Agreed. you might get back the value you wrote, but that is covered by the fact that when reading from a header that was recently made valid, which has not been assigned a value since the header became valid, it might return any bit pattern for the type of that field.  Such a bit pattern might happen to be one that was written to that field at some point in time.